src/app/modules/submission/submission.utils.ts

```diff
@@ -68,3 +92,237 @@ export const extractEmailConfirmationData = (
+      const responseId = r._id
+      const formField = fieldMap[responseId]
+      if (!formField) {
+        return err(
+          new ProcessingError('Response ID does not match form field IDs'),
+        )
+      }
+      return ok(r)
+    })
+
+    const validationResultCombined = combine(validationResultList)
+    if (validationResultCombined.isErr()) {
+      return err(
+        new ProcessingError('Response ID does not match form field IDs'),
+      )
+    }
+
+    return ok(fieldMap as ValidatedFieldMap)
+  }
+
+  /**
+   * Generates a set of response IDs that are visible.
+   * @param responses
+   * @param fieldMap
+   * @param visibilityPredicate - Predicate that determines if a response
+   * is visible.
+   * @returns visibleResponseIds - Generated set of response IDs.
+   * @protected
+   */
+  protected static getVisibleResponseIds(
+    responses: FieldResponse[],
+    fieldMap: ValidatedFieldMap,
+    visibilityPredicate: (r: FieldResponse) => boolean,
+  ): VisibleResponseIdSet {
+    return responses.reduce<FieldIdSet>((acc, response) => {
+      const responseId = response._id
+      if (visibilityPredicate(response)) {
+        acc.add(responseId)
+      }
+      return acc
+    }, new Set()) as VisibleResponseIdSet
+  }
+
+  /**
+   * Generates a set of response IDs that are verifiable. A response is
+   * said to be verifiable if its corresponding form field has the
+   * isVerifiable property set to `true`.
+   * @param responses
+   * @param fieldMap
+   * @returns verifiableResponseIds - Generated set of response IDs.
+   * @protected
+   */
+  protected static getVerifiableResponseIds(
+    responses: FieldResponse[],
+    fieldMap: ValidatedFieldMap,
+  ): VerifiableResponseIdSet {
+    return responses.reduce<FieldIdSet>((acc, response) => {
+      const responseId = response._id
+      const formField = fieldMap[responseId]
+      if (formField.isVerifiable) {
+        acc.add(responseId)
+      }
+      return acc
+    }, new Set()) as VerifiableResponseIdSet
+  }
+
+  /**
+   * Prior to the introduction of IncomingResponse, responses were
+   * stored in a ProcessedFieldResponse[]. This method helps to
+   * create a ProcessedFieldResponse object that can be used in
+   * functions that have not been refactored to take in the new
+   * IncomingResponse object.
+   *
+   * A ProcessedResponse is just a regular FieldResponse with some
+   * metadata inserted into them. This function takes in several
+   * data sets that help determine what each metadata field should
+   * contain.
+   * @param response
+   * @param fieldMap
+   * @param visibleResponseIds
+   * @param verifiableResponseIds
+   * @returns processedFieldResponse - The ProcessedFieldResponse
+   * @protected
+   */
+  protected static getLegacyProcessedFieldResponse(
+    response: FieldResponse,
+    fieldMap: ValidatedFieldMap,
+    visibleResponseIds: VisibleResponseIdSet,
+    verifiableResponseIds: VerifiableResponseIdSet,
```

Comment:
suggest making this an object instead of positional arguments as discussed on slack